tests/unit/test_dbapi_connection.py

```diff
@@ -219,8 +219,14 @@ def test_does_not_keep_cursor_instances_alive(self):
         # Connections should not hold strong references to the Cursor instances
         # they created, unnecessarily keeping them alive.
         gc.collect()
-        cursors = [obj for obj in gc.get_objects() if isinstance(obj, Cursor)]
-        self.assertEqual(len(cursors), 2)
+        cursor_count = 0
+        for obj in gc.get_objects():
+          try:
+            if isinstance(obj, Cursor):
+              cursor_count += 1
+          except ReferenceError:  # pragma: NO COVER
+            pass
+        self.assertEqual(cursor_count, 2)
```

Comment:
Looking at the linked approach TensorFlow took, this looks good to me.

Out of curiosity and because I cannot reproduce the test failure locally (I'm on Linux), I'd just like to better understand what the issue is here.

Specifically, why would the check `isinstance(obj, Cursor)` raise a `ReferenceError` if a weak reference to `obj` does not exist anymore? And it's only an issue with the test, not the `Connection` code itself?

Reply:
Check out [weakrefobject.c](https://github.com/python/cpython/blob/main/Objects/weakrefobject.c#L416-L437).

The weakref `obj` is unwrapped when calling `isinstance` and if the object has been purged, it will fail. So this should be unrelated to `Cursor` or any of the code here but can be triggered by dependent libraries doing funny things with objects (think protos).

Reply:
> The weakref obj is unwrapped when calling isinstance

You mean `weakref.proxy`? Because when initially locally testing with a dead `weakref.weakref`, it simply returned `None` and `isinstance(wref, Foo)` returned `False` (no `ReferenceError`). However, the error is indeed reproducible when using a dead `weakref.proxy` instead.

I suppose `obj` returned by `gc.get_objects()` is actually a proxy, which explains the behavior. Thanks!